design/existing-resource-policy_design.md

```diff
@@ -0,0 +1,247 @@
+
+## Non Goals
+- Add support for `ExistingResourcePolicy` to restore API for Non-Kubernetes resources.
+- Add support for `ExistingResourcePolicy` to restore API for `PersistentVolume` data.
+- Change existing restore workflow for `ServiceAccount` objects
+- Add support for `ExistingResourcePolicy` as `recreate` for Kubernetes resources. (Future scope feature)
+
+### Use-cases/Scenarios
+
+### A. Production Cluster - Backup Cluster:
+Let's say you have a Backup Cluster which is identical to the Production Cluster. After some operations/usage/time the Production Cluster had changed itself, there might be new deployments, some secrets might have been updated. Now, this means that the Backup cluster will no longer be identical to the Production Cluster. In order to keep the Backup Cluster up to date/identical to the Production Cluster with respect to Kubernetes resources except PV data we would like to use Velero for scheduling new backups which would in turn help us update the Backup Cluster via Velero restore.
+
+Reference: https://github.com/vmware-tanzu/velero/issues/4066#issuecomment-954320686
+
+### B. Help identify resource delta:
+Here delta resources mean the resources restored by a previous backup, but they are no longer in the latest backup. Let's follow a sequence of steps to understand this scenario:
+- Consider there are 2 clusters, Cluster A, which has 3 resources - P1, P2 and P3.
+- Create a Backup1 from Cluster A which has P1, P2 and P3.
+- Perform restore on a new Cluster B using Backup1.
+- Now, Lets say in Cluster A resource P1 gets deleted and resource P2 gets updated.
+- Create a new Backup2 with the new state of Cluster A, keep in mind Backup1 has P1, P2 and P3 while Backup2 has P2' and P3.
+- So the Delta here is (|Cluster B - Backup2|), Delete P1 and Update P2.
+- During Restore time we would want the Restore to help us identify this resource delta.
+
+Reference: https://github.com/vmware-tanzu/velero/pull/4613#issuecomment-1027260446
+
+## High-Level Design
+### Approach 1: Add a new spec field `existingResourcePolicy` to the Restore API
+In this approach we do *not* change existing velero behavior. If the resource to restore in cluster is equal to the one backed up then do nothing following current Velero behavior. For resources that already exist in the cluster that are not equal to the resource in the backup (other than Service Accounts). We add a new optional spec field `existingResourcePolicy` which can have the following values:
+1. `none`: This is the existing behavior, if Velero encounters a resource that already exists in the cluster, we simply
+skip restoration.
+2. `update` or `merge` or `patch`: In this behaviour Velero will try an attempt to patch the resource with the backed up version, 
+if the patch fails log it as a warning and continue the restore process, user may wish specify the merge type.
+3. `updateAll`: Similar to `update` option but Velero will update labels on unchanged resources too.
```

Comment:
Thoughts?
```suggestion
3. `updateOverrideLabels`: Similar to `update` option but Velero will update labels on unchanged resources too.
```

Reply:
```suggestion
3. `updateForceUnchangedLabels`: Similar to `update` option but Velero will update labels on unchanged resources too.
```

Reply:
It's not so much "unchanged" labels we're forcing, it's the labels that *velero* adds on restore -- these labels aren't in the backup. `updateForceVeleroLabels` is more accurate, but I think it's a bit clumsy as a config option since it's so long. Perhaps just `updateVeleroLabels` -- that's less clumsy, but it might confuse users into thinking we're *only* updating velero labels. I'm thinking that `updateAll` may be best, as long as the documentation is clear on what this means: "Update all resources, including unchanged ones, in order to ensure that the velero labels related to the backup and restore are appropriately updated."

Reply:
When was the last time verbosity caused problems? It accelerated java adoption. You'd most likely type it once in Jason/yaml or velero cli autocompletes or also have an accompanying shorthand flag.

Reply:
I prefer to cater flags to "I haven't read the doc" folks more than the "I type this out everyday so it has to be short".. since support escalations come from the former.

Reply:
`updateIncludingVeleroLabelsForAll` would probably be the most accurate. Or perhaps `updateAndForceVeleroLabels`. I'd still vote for `updateAll`, but if we want something longer, one of these would be clearer than the others I think.